Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -166,10 +172,22 @@ def nest : DyckWord where
 @[simp] lemma nest_ne_zero : p.nest ≠ 0 := by simp [← toList_ne_nil, nest]
 
 variable (p) in
-/-- Denest `p`, i.e. `(x)` becomes `x`, given that `p` is strictly positive in its interior
-(this ensures that `x` is a Dyck word). -/
-def denest (pos : ∀ i, 0 < i → i < p.toList.length →
-    (p.toList.take i).count D < (p.toList.take i).count U) : DyckWord where
+/-- A property stating that `p` is strictly positive in its interior, i.e. is of the form `(x)`
+with `x` a Dyck word. -/
+def IsNested : Prop :=
+  ∀ i, 0 < i → i < p.toList.length → (p.toList.take i).count D < (p.toList.take i).count U
+
+lemma isNested_nest : p.nest.IsNested := fun i lb ub ↦ by
+  simp_rw [nest, length_append, length_singleton] at ub ⊢
+  rw [take_append_of_le_length (by rw [singleton_append, length_cons]; omega),
+    take_append_eq_append_take, take_of_length_le (by rw [length_singleton]; omega),
+    length_singleton, singleton_append, count_cons_of_ne (by simp), count_cons_self,
+    Nat.lt_add_one_iff]
+  exact p.count_D_le_count_U _
+
+variable (p) in
+/-- Denest `p`, i.e. `(x)` becomes `x`, given that `p.IsNested`. -/
+def denest (hn : p.IsNested) : DyckWord where
```

Comment:
Can you either make `denest` return `0` on the empty word or include `p ≠ 0` in `IsNested`?

Reply:
I've gone for the second option.